File: Greedy/452.minimum-number-of-arrows-to-burst-balloons.py
#
# @lc app=leetcode id=452 lang=python3
#
# [452] Minimum Number of Arrows to Burst Balloons
#

# @lc code=start
class Solution:
    def findMinArrowShots(self, points: List[List[int]]) -> int:
        res = 1
        points = sorted(points)
        print(points)
        prev_end = points[0][1]

        for start, end in points[1:]:
            if start > prev_end:
                res += 1
                prev_end = end
            else:
                prev_end = min(prev_end, end)

        return res
# @lc code=end

File: Interval/452.minimum-number-of-arrows-to-burst-balloons.java

/*
 * @lc app=leetcode id=452 lang=java
 *
 * [452] Minimum Number of Arrows to Burst Balloons
 */
import java.util.Arrays;

// @lc code=start
class Solution {
    public int findMinArrowShots(int[][] points) {
        Arrays.sort(points, (a, b) -> Integer.compare(a[0], b[0]));
        int res = 1;
        int right = points[0][1];

        for (int i = 1; i < points.length; i++) {
            if (points[i][0] <= right) {
                right = Math.min(points[i][1], right);
            } else {
                res += 1;
                right = points[i][1];
            }
        }
        return res;
    }
}
// @lc code=end
